packages/create-cloudflare/src/templateMap.ts

```diff
@@ -1,53 +1,368 @@
-				...args,
-				ts: true,
-			}),
-	},
-	openapi: {
-		label: `OpenAPI 3.1`,
-		handler: (args) =>
-			runWorkersGenerator({
-				...args,
-				ts: true,
-			}),
-	},
-	"pre-existing": {
-		label: "Pre-existing Worker (from Dashboard)",
-		handler: runWorkersGenerator,
-		hidden: true,
-	},
+// A template can have a number of variants, usually js/ts
+type VariantInfo = {
+	path: string;
+};
+
+type StaticFileMap = Record<string, VariantInfo>;
+
+export type FrameworkMap = Awaited<ReturnType<typeof getFrameworkMap>>;
+export type FrameworkName = keyof FrameworkMap;
+
+export const getFrameworkMap = async () => ({
+	angular: (await import("../templates/angular/c3")).default,
+	astro: (await import("../templates/astro/c3")).default,
+	docusaurus: (await import("../templates/docusaurus/c3")).default,
+	gatsby: (await import("../templates/gatsby/c3")).default,
+	hono: (await import("../templates/hono/c3")).default,
+	next: (await import("../templates/next/c3")).default,
+	nuxt: (await import("../templates/nuxt/c3")).default,
+	qwik: (await import("../templates/qwik/c3")).default,
+	react: (await import("../templates/react/c3")).default,
+	remix: (await import("../templates/remix/c3")).default,
+	solid: (await import("../templates/solid/c3")).default,
+	svelte: (await import("../templates/svelte/c3")).default,
+	vue: (await import("../templates/vue/c3")).default,
+});
+
+export const getTemplateMap = async () => {
+	return {
+		"hello-world": await import("../templates/hello-world/c3.json"),
+		// Dummy record
+		webFramework: { displayName: "Website or web app" },
+		common: await import("../templates/common/c3.json"),
+		scheduled: await import("../templates/scheduled/c3.json"),
+		queues: await import("../templates/queues/c3.json"),
+		chatgptPlugin: await import("../templates/chatgptPlugin/c3.json"),
+		openapi: await import("../templates/openapi/c3.json"),
+		"remote-template": { displayName: "From a remote git repository" },
+		"pre-existing": (await import("../templates/pre-existing/c3")).default,
+	} as unknown as Record<string, TemplateConfig>;
 };
+
+export const selectTemplate = async (args: Partial<C3Args>) => {
+	// If not specified, attempt to infer the `type` argument from other flags
+	if (!args.type) {
+		if (args.framework) {
+			args.type = "webFramework";
+		} else if (args.existingScript) {
+			args.type = "pre-existing";
+		} else if (args.template) {
+			args.type = "remote-template";
+		}
+	}
+
+	const templateMap = await getTemplateMap();
+
+	const templateOptions = Object.entries(templateMap).map(
+		([value, { displayName, hidden }]) => ({
+			value,
+			label: displayName,
+			hidden,
+		})
+	);
+
+	const type = await processArgument<string>(args, "type", {
+		type: "select",
+		question: "What type of application do you want to create?",
+		label: "type",
+		options: templateOptions,
+		defaultValue: C3_DEFAULTS.type,
+	});
+
+	if (!type) {
+		return crash("An application type must be specified to continue.");
+	}
```

Comment:
can the type really not be defined here given the `processArgument` call above? (which has a `defaultValue`)

are you're trying to be extra defensive with this check? 🤔

Reply:
PS: does `processArgument` allow users to provide an empty input? 🤔 (if so, is that correct?)

Reply:
It guards against this one specific silly use case: `npm create cloudflare -- --type`.

`processArgument` won't allow users to provide empty input interactively, but it will skip interactive prompting if a user has specific command line input. And if they happen to supply empty string, this crash will occur.

Reply:
ahhh ok yeah, I think I remember something like this! thanks for the explanation! :slightly_smiling_face: